File: SAP/sap.ml
(* simple as possible file system: (inefficient, only for small devices)
 * list of block groups (BG), each of which is a file or empty spaces.
 * a BG begins with a 64-bit/8-byte size in bytes, then the name, then the data
 * the size includes the initial 8 bytes and the null-terminated name
 * for an empty space, the name is empty, meaning only the null byte. *)

 type device = { 
  name: string; 
  bs: int64; 
  nb: int64;
  r: int64 -> int list; 
  w: int64 -> int list -> unit }

(* 1. assert: every write operation in this program writes device.bs bytes,
*            every read operation after calling format reads device.bs bytes *)
(* 2. assert: the name of every file fits in device.bs - 9 (see new_file) *)
(* 3. assert: unless there are errors, every file written is accessible
until a matching delete_file (or forever). *)

let check_invariant1 bs lst = assert (Int64.of_int (List.length lst) = bs)
let check_invariant2 bs name =
assert (List.length name <= (Int64.to_int bs) - 9)

(* generic functions for groups of bytes *)
let rec getn lst n =    (* get n bytes from the list *)
if n <= 0 then []
else match lst with
| [] -> failwith "insufficient bytes"
| first :: rest -> first :: (getn rest (n - 1))

let rec restn lst n =   (* remove first n bytes from the list *)
if n <= 0 then lst
else match lst with
| [] -> failwith "unable to skip over n bytes"
| _ :: rest -> restn rest (n - 1)

(* convert a 64-bit number to a list of bytes and get from a list of bytes *)
let n64bytes n =
let rec make_rev_bytes x index =
if index <= 0 then []
else (Int64.to_int (Int64.rem x 256L)) ::
make_rev_bytes (Int64.div x 256L) (index - 1) in
List.rev (make_rev_bytes n 8)

let bytes64 lst =
let first8 = getn lst 8 in
let rec compute lst =    (* lst has the lsb first *)
match lst with
| [] -> 0L
| first :: rest ->
(assert (first >= 0); assert (first < 256);
(Int64.add (Int64.of_int first)
          (Int64.mul (compute rest) 256L))) in
compute (List.rev first8)

(* 4. assert bytes64 (n64bytes n) = n for any n < 2^63 *)
(* 5. assert n64bytes (bytes64 l) = l for any list of int<256 of length = 8 *)
let check_invariant45 () =
let n = Random.int64 2147483646L in
let rec make_list = function
| 0 -> []
| l -> (Random.int 256) :: (make_list (l - 1)) in
let make_first_less_than_128 = function
| [] -> failwith "illegal list does not have first"
| first :: rest ->
  (if first >= 128 then first - 128 else first) :: rest in
let rec list_to_string = function
| [] -> ""
| first :: rest ->
 (string_of_int first) ^ "." ^ (list_to_string rest) in
let random_list = make_first_less_than_128 (make_list 8) in
(print_string "random int is " ; print_endline (Int64.to_string n);
print_string "random list is " ;
print_endline (list_to_string random_list);
print_string "random list is now " ;
print_endline (list_to_string (n64bytes (bytes64 random_list)));
assert (bytes64 (n64bytes n) = n);
assert (n64bytes (bytes64 random_list) = random_list))

let _ = Random.self_init()
let _ = check_invariant45();;

(* make a list of n zeros *)
let rec make_zeros n =
if n = 0L then []
else 0 :: (make_zeros (Int64.sub n 1L))

(* create a block marking an empty set of nb blocks or bytes *)
let rec make_empty_bytes device nb = n64bytes nb @
                      make_zeros (Int64.sub device.bs 8L)
                      
let rec make_empty_block device nb =
make_empty_bytes device (Int64.mul nb device.bs)

let blocks_for_bytes device n =
Int64.div (Int64.add n (Int64.sub device.bs 1L)) device.bs

(* clear the device and create an empty set of all blocks *)
let format device =
let _ = assert (device.bs > 9L) in
let zeros = make_zeros device.bs in
let rec write_all n =
if n >= device.nb then ()
else (device.w n zeros; write_all (Int64.add n 1L)) in
(write_all 1L;
device.w 0L (make_empty_block device device.nb))

(* find an unused block with at least size bytes *)
let find_empty device size =  (* size includes 8+namelen+1 bytes of header *)
let blocks = blocks_for_bytes device size in
let rec helper bn =
if bn >= device.nb then (-1L, 0L)       (* not found *)
else
 (let block = device.r bn in
  let fsize = bytes64 block in       (* size in bytes *)
  let nb = blocks_for_bytes device fsize in (* size in blocks *)
  let name = restn block 8 in
  if nb >= blocks && getn name 1 = [0] then  (* empty block *)
    (if nb > blocks then             (* some free space *)
       let num_empty = Int64.sub nb blocks in
       let new_empty = make_empty_block device num_empty
       and new_bn = Int64.add bn blocks in
       device.w new_bn new_empty
     else ();
     (bn, nb))
  else                            (* go to next block *)
    helper (Int64.add bn nb)) in
helper 0L

(* used in the invariants checking *)
let read_errors = ref 0
let write_errors = ref 0
let files_created = ref ([]: (string * int list) list)
let remove_file_from_created name =
let rec remove_file = function
| [] -> []  (* fail silently if the file doesn't exist *)
| (fname, fcontent) :: rest ->
 if fname = name then rest
 else (fname, fcontent) :: (remove_file rest) in
files_created := remove_file (! files_created)

let rec compare_byte_lists n c1 c2 = match (c1, c2, n) with
| ([], [], _) -> true
| (_ :: _, [], None) -> false
| (_ :: _, [], Some _) -> print_endline "c1 is longer"; false
| ([], _ :: _, None) -> false
| ([], _ :: _, Some _) -> print_endline "c2 is longer"; false
| (h1 :: r1, h2 :: r2, None) ->
h1 = h2 && compare_byte_lists None r1 r2
| (h1 :: r1, h2 :: r2, Some elt) ->
(if h1 != h2 then
 (print_int elt; print_string ": ";
  print_int h1; print_string " != ";
  print_int h2; print_endline ""; false)
else compare_byte_lists (Some (elt + 1)) r1 r2)

let internal_get_header device bn =
let block = device.r bn in
let fsize = bytes64 block in       (* size in bytes *)
let nb = blocks_for_bytes device fsize in (* size in blocks *)
let fname_plus = restn block 8 in
let rec separate_name = function
| [] -> ([], [])
| 0 :: rest -> ([0], rest)
| c :: rest ->
 let (name_end, list) = separate_name rest in
 (c :: name_end, list) in
let (fname, initial_data) = separate_name fname_plus in
check_invariant2 device.bs fname;
(fsize, nb, fname, initial_data, block)

let blocks_read = ref []

(* read or delete are almost the same, so do both in the same function *)
let internal_read_or_delete do_delete device name =

let rec name_to_list pos =
if pos = String.length name then [0]
else (Char.code (String.get name pos)) ::
    name_to_list (pos + 1) in

let lname = name_to_list 0 in

let rec clear_file bn nb zero_block =
if nb <= 0L then ()
else (device.w bn zero_block;
     clear_file (Int64.add bn 1L) (Int64.sub nb 1L) zero_block) in

let rec delete_file bn nb prev_bn prev_nb =
let _ = remove_file_from_created name in
let get_next bn =
    if bn >= device.nb then 0L
    else (let (_, nb, fname, _, _) =
               internal_get_header device bn in
          if List.length fname = 1 then
            (device.w bn (make_zeros device.bs);
             nb)
          else 0L) in
let next_nb = get_next (Int64.add bn nb) in
let size_from_bn = Int64.mul (Int64.add nb next_nb) device.bs in
let max_bytes = Int64.add size_from_bn
                         (Int64.mul prev_nb device.bs) in
(clear_file bn nb (make_zeros device.bs);
if prev_bn = -1L then
  device.w bn (make_empty_bytes device size_from_bn)
else
  device.w prev_bn (make_empty_bytes device max_bytes)) in

let rec read_all block_num remaining_size =
let block = device.r block_num in
if remaining_size > device.bs then
  block @ (read_all (Int64.add block_num 1L)
                    (Int64.sub remaining_size device.bs))
else getn block (Int64.to_int remaining_size) in


let rec find_file bn prev_bn prev_nb =
if bn >= device.nb then None          (* not found *)
else
 (let (fsize, nb, fname, initial_data, block) =
        internal_get_header device bn in
  let hsize = Int64.of_int (8 + List.length fname) in
  let dsize = Int64.sub fsize hsize in
  let next_bn = Int64.add bn 1L in
  if compare_byte_lists None lname fname then       (* found *)
    (let result =
       if fsize <= device.bs then getn initial_data
                                       (Int64.to_int dsize)
       else initial_data @
            (read_all next_bn (Int64.sub fsize device.bs)) in
     (if do_delete then delete_file bn nb prev_bn prev_nb
      else ();
      Some result))
  else                               (* not found *)
    (if List.length fname = 1 then   (* empty, possible merge *)
       find_file (Int64.add bn nb) bn nb
     else find_file (Int64.add bn nb) (-1L) (-1L))) in
     
let result = find_file 0L (-1L) (-1L) in

result

let check_invariant3 dev =
if ! read_errors = 0 && ! write_errors = 0 then
(let rec verify_files = function
| [] -> true
| (name, content) :: rest ->
 (match internal_read_or_delete false dev name with
  | None -> print_endline (name ^ " not found"); false
  | Some c ->
      if compare_byte_lists (Some 0) c content
      then verify_files rest
      else (print_endline (name ^ " not same");
            print_int (List.length c);
            print_string " =? ";
            print_int (List.length content);
            print_endline "";
            false)) in
assert (verify_files (! files_created)))
else ()

let read_file device name =
check_invariant3 device;
internal_read_or_delete false device name

let delete_file device name =
check_invariant3 device;
internal_read_or_delete true device name

(* create a new file with the given data *)
let new_file device name data =

let _ = check_invariant3 device in
if read_file device name = None &&
(Int64.of_int (String.length name + 9)) < device.bs then
(let rec name_to_list pos =
   if pos = String.length name then [0]
   else (Char.code (String.get name pos)) ::
        name_to_list (pos + 1) in

let lname = name_to_list 0 in

let rec write_data bn data dlen =
   if dlen <= device.bs then
     device.w bn (data @ (make_zeros (Int64.sub device.bs dlen)))
   else
     let first = getn data (Int64.to_int device.bs)
     and rest = restn data (Int64.to_int device.bs) in
     (device.w bn first;
      write_data (Int64.add bn 1L) rest
                 (Int64.sub dlen device.bs)) in

let size = Int64.of_int (8 + (List.length lname) + (List.length data)) in
let all_data = (n64bytes size) @ lname @ data in
let (bn, nb) = find_empty device size in
if bn < 0L then failwith "unable to create file"
else files_created := (name, data) :: (! files_created);
write_data bn all_data size)

else if Int64.of_int (String.length name + 9) >= device.bs then
failwith "new_file name too long"

else
failwith ("file " ^ name ^ " already exists")

let list_files device =
let list_to_name name =
let int_to_string i = String.make 1 (Char.chr i) in
let substrings = List.map int_to_string name in
String.concat "" substrings in
let rec helper bn =
if bn >= device.nb then []          (* no more files *)
else
let (fsize, nb, fname, initial_data, block) =
   internal_get_header device bn in
let name = list_to_name (getn fname (List.length fname - 1)) in
let hsize = Int64.of_int (8 + List.length fname) in
let dsize = Int64.sub fsize hsize in
let others = helper (Int64.add bn nb) in
if List.length fname > 1 then (name, bn, dsize) :: others
                      else others in
check_invariant3 device;
helper 0L

(* test stuff *)
let blocks = Array.make 500 []
let read_block bs n =
let data = Array.get blocks (Int64.to_int n) in
check_invariant1 bs data;
data

(* let write_block n b = Array.set blocks n b *)
let write_block bs n b =
check_invariant1 bs b;
Array.set blocks (Int64.to_int n) b
let dev = let bs = 512L in
{ name = "test"; bs = bs; nb = 100L; r = read_block bs; w = write_block bs }
let rec make_data = function
| 0 -> []
| n -> (n mod 256) :: make_data (n - 1)
let data = [1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 11; 12]
(* two blocks' worth of data *)
let ld = data @ data @ data @ data @ data @ data @ data @ data @ data @ data @
data @ data @ data @ data @ data @ data @ data @ data @ data @ data @
data @ data @ data @ data @ data @ data @ data @ data @ data @ data @
data @ data @ data @ data @ data @ data @ data @ data @ data @ data @
data @ data @ data @ data @ data @ data @ data @ data @ data @ data @
data @ data @ data @ data @ data @ data @ data @ data @ data @ data @
data @ data @ data @ data @ data @ data @ data @ data @ data @ data

let rec make_files device = function
| [] -> ()
| (fname, fsize) :: rest ->
(print_endline ("making file " ^ fname);
new_file device fname (make_data fsize);
make_files device rest)
let sample_files1 = [("a", 10); ("b", 600); ("c", 512); ("d", 500); ("e", 501);
      ("f", 502); ("g", 503); ("h", 504); ("i", 505);
      ("j", 10000); ("k", 1); ("l", 0)]

(* print the non-zero part of non-zero blocks *)
let print_blocks device max =
let blen = Int64.of_int (Array.length blocks) in
let dlen = if device.nb < blen then device.nb else blen in
let actual_max = if max <= 0L || max > dlen then dlen else max in
let rec is_zeros = function
| [] -> true
| 0 :: rest -> is_zeros rest
| _ -> false in
let rec print_block lst min =
if min <= 0L && is_zeros lst then ()
else match lst with
    | [] -> if min <= 0L then failwith "no bytes in print_block"
            else print_block (make_zeros min) min
    | first :: rest ->
       print_int first;
       print_string " ";
       print_block rest (Int64.sub min 1L) in
let rec print_all n =
if n < actual_max then
 (let a = read_block device.bs n in
  if is_zeros a then ()
  else
    (print_int (Int64.to_int n);
     print_string ": ";
     print_block a 8L;
     print_endline "");
  print_all (Int64.add n 1L))
else () in
print_all 0L

let p () = print_blocks dev 0L

File: SAP/sap.hs

{-
  Simple as possible file system: (inefficient, only for small devices)
  List of block groups (BG), each of which is a file or empty spaces.
  A BG begins with a 64-bit/8-byte size in bytes, then the name, then the data
  The size includes the initial 8 bytes and the null-terminated name
  For an empty space, the name is empty, meaning only the null byte.
-}

module SimpleFS where

import Data.Word
import Data.Int
import Data.Bits
import Data.Char
import Data.List (intercalate)
import Control.Monad (when)
import System.Random
import qualified Data.Array as Array

-- Device type definition
data Device = Device { 
  name :: String,
  bs :: Int64,
  nb :: Int64,
  r :: Int64 -> [Int],
  w :: Int64 -> [Int] -> IO ()
}

{-
  1. assert: every write operation in this program writes device.bs bytes,
     every read operation after calling format reads device.bs bytes
  2. assert: the name of every file fits in device.bs - 9 (see new_file)
  3. assert: unless there are errors, every file written is accessible
     until a matching delete_file (or forever).
-}

checkInvariant1 :: Int64 -> [Int] -> IO ()
checkInvariant1 bs lst = 
  if fromIntegral (length lst) == bs 
  then return () 
  else error "Invariant 1 violated"

checkInvariant2 :: Int64 -> [Int] -> IO ()
checkInvariant2 bs name =
  if length name <= fromIntegral bs - 9
  then return ()
  else error "Invariant 2 violated"

-- Generic functions for groups of bytes
getn :: [Int] -> Int -> [Int]  -- get n bytes from the list
getn lst n
  | n <= 0    = []
  | otherwise = case lst of
      []           -> error "insufficient bytes"
      first : rest -> first : getn rest (n - 1)

restn :: [Int] -> Int -> [Int]  -- remove first n bytes from the list
restn lst n
  | n <= 0    = lst
  | otherwise = case lst of
      []          -> error "unable to skip over n bytes"
      _ : rest    -> restn rest (n - 1)

-- Convert a 64-bit number to a list of bytes and get from a list of bytes
n64bytes :: Int64 -> [Int]
n64bytes n = reverse $ makeRevBytes n 8
  where
    makeRevBytes :: Int64 -> Int -> [Int]
    makeRevBytes x index
      | index <= 0 = []
      | otherwise  = fromIntegral (x `mod` 256) : makeRevBytes (x `div` 256) (index - 1)

bytes64 :: [Int] -> Int64
bytes64 lst = compute $ reverse $ getn lst 8
  where
    compute :: [Int] -> Int64
    compute []          = 0
    compute (first:rest) = do
      if first < 0 || first >= 256
        then error "Byte value out of range"
        else fromIntegral first + 256 * compute rest

{-
  4. assert bytes64 (n64bytes n) = n for any n < 2^63
  5. assert n64bytes (bytes64 l) = l for any list of int<256 of length = 8
-}
checkInvariant45 :: IO ()
checkInvariant45 = do
  let maxInt = 2147483646 :: Int64
  n <- randomRIO (0, maxInt)
  
  let makeList 0 = []
      makeList l = randomRIO (0, 255) >>= \r -> (r :) <$> makeList (l - 1)
  
  let makeFirstLessThan128 [] = error "illegal list does not have first"
      makeFirstLessThan128 (first:rest) = 
        (if first >= 128 then first - 128 else first) : rest
  
  let listToString [] = ""
      listToString (first:rest) = show first ++ "." ++ listToString rest
  
  randomList <- makeList 8 >>= return . makeFirstLessThan128
  
  putStr "random int is "
  putStrLn $ show n
  putStr "random list is "
  putStrLn $ listToString randomList
  putStr "random list is now "
  putStrLn $ listToString (n64bytes (bytes64 randomList))
  
  when (bytes64 (n64bytes n) /= n) $
    error "Invariant 4 violated: bytes64 (n64bytes n) != n"
  
  when (n64bytes (bytes64 randomList) /= randomList) $
    error "Invariant 5 violated: n64bytes (bytes64 randomList) != randomList"

-- Make a list of n zeros
makeZeros :: Int64 -> [Int]
makeZeros n
  | n == 0    = []
  | otherwise = 0 : makeZeros (n - 1)

-- Create a block marking an empty set of nb blocks or bytes
makeEmptyBytes :: Device -> Int64 -> [Int]
makeEmptyBytes device nb = n64bytes nb ++ makeZeros (bs device - 8)

makeEmptyBlock :: Device -> Int64 -> [Int]
makeEmptyBlock device nb = makeEmptyBytes device (nb * bs device)

blocksForBytes :: Device -> Int64 -> Int64
blocksForBytes device n = (n + bs device - 1) `div` bs device

-- Clear the device and create an empty set of all blocks
format :: Device -> IO ()
format device = do
  when (bs device <= 9) $ error "Block size must be > 9"
  let zeros = makeZeros (bs device)
  let writeAll n
        | n >= nb device = return ()
        | otherwise = do
            w device n zeros
            writeAll (n + 1)
  
  writeAll 1
  w device 0 (makeEmptyBlock device (nb device))

-- Find an unused block with at least size bytes
findEmpty :: Device -> Int64 -> IO (Int64, Int64)
findEmpty device size = do  -- size includes 8+namelen+1 bytes of header
  let blocks = blocksForBytes device size
  let helper bn
        | bn >= nb device = return (-1, 0)  -- not found
        | otherwise = do
            let block = r device bn
            let fsize = bytes64 block           -- size in bytes
            let nb = blocksForBytes device fsize  -- size in blocks
            let name = restn block 8
            if nb >= blocks && getn name 1 == [0] then do  -- empty block
              when (nb > blocks) $ do             -- some free space
                let numEmpty = nb - blocks
                let newEmpty = makeEmptyBlock device numEmpty
                let newBn = bn + blocks
                w device newBn newEmpty
              return (bn, nb)
            else                            -- go to next block
              helper (bn + nb)
  helper 0

-- Used in the invariants checking
type FileCreated = (String, [Int])
type FileContent = (String, [Int])

newIORef :: a -> IO (IORef a)
newIORef = undefined  -- This would be implemented with IORef in real code

readIORef :: IORef a -> IO a
readIORef = undefined  -- This would be implemented with IORef in real code

writeIORef :: IORef a -> a -> IO ()
writeIORef = undefined  -- This would be implemented with IORef in real code

modifyIORef :: IORef a -> (a -> a) -> IO ()
modifyIORef = undefined  -- This would be implemented with IORef in real code

readErrors :: IORef Int
readErrors = undefined  -- This would be initialized to 0

writeErrors :: IORef Int
writeErrors = undefined  -- This would be initialized to 0

filesCreated :: IORef [FileCreated]
filesCreated = undefined  -- This would be initialized to []

removeFileFromCreated :: String -> IO ()
removeFileFromCreated name = modifyIORef filesCreated removeFile
  where
    removeFile [] = []  -- fail silently if the file doesn't exist
    removeFile ((fname, fcontent):rest)
      | fname == name = rest
      | otherwise     = (fname, fcontent) : removeFile rest

compareByteLists :: Maybe Int -> [Int] -> [Int] -> IO Bool
compareByteLists n c1 c2 = case (c1, c2, n) of
  ([], [], _) -> return True
  (_ : _, [], Nothing) -> return False
  (_ : _, [], Just _) -> putStrLn "c1 is longer" >> return False
  ([], _ : _, Nothing) -> return False
  ([], _ : _, Just _) -> putStrLn "c2 is longer" >> return False
  (h1 : r1, h2 : r2, Nothing) ->
    if h1 == h2 
    then compareByteLists Nothing r1 r2
    else return False
  (h1 : r1, h2 : r2, Just elt) ->
    if h1 == h2 
    then compareByteLists (Just (elt + 1)) r1 r2
    else do
      putStr $ show elt ++ ": "
      putStr $ show h1 ++ " != "
      putStr $ show h2 ++ "\n"
      return False

internalGetHeader :: Device -> Int64 -> IO (Int64, Int64, [Int], [Int], [Int])
internalGetHeader device bn = do
  let block = r device bn
  let fsize = bytes64 block           -- size in bytes
  let nb = blocksForBytes device fsize  -- size in blocks
  let fnameplus = restn block 8
  let separateName [] = ([], [])
      separateName (0:rest) = ([0], rest)
      separateName (c:rest) = 
        let (nameEnd, list) = separateName rest
        in (c : nameEnd, list)
  let (fname, initialData) = separateName fnameplus
  checkInvariant2 (bs device) fname
  return (fsize, nb, fname, initialData, block)

blocksRead :: IORef [Int64]
blocksRead = undefined  -- This would be initialized to []

-- Read or delete are almost the same, so do both in the same function
internalReadOrDelete :: Bool -> Device -> String -> IO (Maybe [Int])
internalReadOrDelete doDelete device name = do
  let nameToList pos
        | pos == length name = [0]
        | otherwise = ord (name !! pos) : nameToList (pos + 1)
  
  let lname = nameToList 0
  
  let clearFile bn nb zeroBlock
        | nb <= 0 = return ()
        | otherwise = do
            w device bn zeroBlock
            clearFile (bn + 1) (nb - 1) zeroBlock
  
  let deleteFile bn nb prevBn prevNb = do
      removeFileFromCreated name
      let getNext bn'
            | bn' >= nb device = return 0
            | otherwise = do
                (_, nb'', fname, _, _) <- internalGetHeader device bn'
                if length fname == 1
                  then do
                    w device bn' (makeZeros (bs device))
                    return nb''
                  else return 0
      nextNb <- getNext (bn + nb)
      let sizeFromBn = (nb + nextNb) * bs device
      let maxBytes = sizeFromBn + prevNb * bs device
      clearFile bn nb (makeZeros (bs device))
      if prevBn == -1
        then w device bn (makeEmptyBytes device sizeFromBn)
        else w device prevBn (makeEmptyBytes device maxBytes)
      return ()
  
  let readAll blockNum remainingSize
        | remainingSize <= 0 = return []
        | otherwise = do
            let block = r device blockNum
            if remainingSize > bs device
              then do
                rest <- readAll (blockNum + 1) (remainingSize - bs device)
                return $ block ++ rest
              else return $ getn block (fromIntegral remainingSize)
  
  let findFile bn prevBn prevNb
        | bn >= nb device = return Nothing          -- not found
        | otherwise = do
            (fsize, nb', fname, initialData, block) <- internalGetHeader device bn
            let hsize = fromIntegral (8 + length fname)
            let dsize = fsize - hsize
            let nextBn = bn + 1
            equal <- compareByteLists Nothing lname fname
            if equal then do      -- found
              result <- if fsize <= bs device
                         then return $ getn initialData (fromIntegral dsize)
                         else do
                           rest <- readAll nextBn (fsize - bs device)
                           return $ initialData ++ rest
              when doDelete $ deleteFile bn nb' prevBn prevNb
              return $ Just result
            else do                               -- not found
              if length fname == 1
                then findFile (bn + nb') bn nb'  -- empty, possible merge
                else findFile (bn + nb') (-1) (-1)
  
  findFile 0 (-1) (-1)

checkInvariant3 :: Device -> IO ()
checkInvariant3 dev = do
  readErrs <- readIORef readErrors
  writeErrs <- readIORef writeErrors
  when (readErrs == 0 && writeErrs == 0) $ do
    created <- readIORef filesCreated
    let verifyFiles [] = return True
        verifyFiles ((name, content):rest) = do
          result <- internalReadOrDelete False dev name
          case result of
            Nothing -> do
              putStrLn (name ++ " not found")
              return False
            Just c -> do
              same <- compareByteLists (Just 0) c content
              if same
                then verifyFiles rest
                else do
                  putStrLn (name ++ " not same")
                  putStr $ show (length c)
                  putStr " =? "
                  putStrLn $ show (length content)
                  return False
    
    verified <- verifyFiles created
    when (not verified) $ error "Invariant 3 violated"

readFile :: Device -> String -> IO (Maybe [Int])
readFile device name = do
  checkInvariant3 device
  internalReadOrDelete False device name

deleteFile :: Device -> String -> IO (Maybe [Int])
deleteFile device name = do
  checkInvariant3 device
  internalReadOrDelete True device name

-- Create a new file with the given data
newFile :: Device -> String -> [Int] -> IO ()
newFile device name data' = do
  checkInvariant3 device
  fileExists <- readFile device name
  nameLength <- return $ fromIntegral (length name + 9)
  
  case fileExists of
    Just _ -> error ("file " ++ name ++ " already exists")
    Nothing
      | nameLength >= bs device -> error "new_file name too long"
      | otherwise -> do
          let nameToList pos
                | pos == length name = [0]
                | otherwise = ord (name !! pos) : nameToList (pos + 1)
          
          let lname = nameToList 0
          
          let writeData bn ddata dlen
                | dlen <= bs device = 
                    w device bn (ddata ++ makeZeros (bs device - dlen))
                | otherwise = do
                    let first = getn ddata (fromIntegral (bs device))
                    let rest = restn ddata (fromIntegral (bs device))
                    w device bn first
                    writeData (bn + 1) rest (dlen - bs device)
          
          let size = fromIntegral (8 + length lname + length data')
          let allData = n64bytes size ++ lname ++ data'
          (bn, nb) <- findEmpty device size
          when (bn < 0) $ error "unable to create file"
          
          modifyIORef filesCreated ((name, data') :)
          writeData bn allData size

listFiles :: Device -> IO [(String, Int64, Int64)]
listFiles device = do
  checkInvariant3 device
  
  let listToName name =
        let intToString i = [chr i]
            substrings = map intToString name
        in concat substrings
  
  let helper bn
        | bn >= nb device = return []          -- no more files
        | otherwise = do
            (fsize, nb', fname, initialData, block) <- internalGetHeader device bn
            let name = listToName (getn fname (length fname - 1))
            let hsize = fromIntegral (8 + length fname)
            let dsize = fsize - hsize
            others <- helper (bn + nb')
            if length fname > 1
              then return $ (name, bn, dsize) : others
              else return others
  
  helper 0

-- Test stuff
blocks :: Array.Array Int [Int]
blocks = Array.array (0, 499) [(i, []) | i <- [0..499]]

readBlock :: Int64 -> Int64 -> [Int]
readBlock bs n = do
  let data' = blocks Array.! fromIntegral n
  checkInvariant1 bs data'
  data'

writeBlock :: Int64 -> Int64 -> [Int] -> IO ()
writeBlock bs n b = do
  checkInvariant1 bs b
  -- Array.// would update and return a new array, but we're mimicking mutable behavior
  -- In a real implementation, this would use a mutable array or IOArray

makeData :: Int -> [Int]
makeData 0 = []
makeData n = (n `mod` 256) : makeData (n - 1)

testDevice :: Device
testDevice = 
  let bs' = 512
  in Device {
    name = "test",
    bs = bs',
    nb = 100,
    r = readBlock bs',
    w = writeBlock bs'
  }

data' :: [Int]
data' = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

-- Two blocks' worth of data
ld :: [Int]
ld = concat $ replicate 70 data'

makeFiles :: Device -> [(String, Int)] -> IO ()
makeFiles _ [] = return ()
makeFiles device ((fname, fsize):rest) = do
  putStrLn ("making file " ++ fname)
  newFile device fname (makeData fsize)
  makeFiles device rest

sampleFiles1 :: [(String, Int)]
sampleFiles1 = [
  ("a", 10), ("b", 600), ("c", 512), ("d", 500), ("e", 501),
  ("f", 502), ("g", 503), ("h", 504), ("i", 505),
  ("j", 10000), ("k", 1), ("l", 0)]

-- Print the non-zero part of non-zero blocks
printBlocks :: Device -> Int64 -> IO ()
printBlocks device max' = do
  let blen = fromIntegral (Array.bounds blocks) -- this approximates Array.length
  let dlen = if nb device < blen then nb device else blen
  let actualMax = if max' <= 0 || max' > dlen then dlen else max'
  
  let isZeros [] = True
      isZeros (0:rest) = isZeros rest
      isZeros _ = False
  
  let printBlock lst min'
        | min' <= 0 && isZeros lst = return ()
        | otherwise = case lst of
            [] -> if min' <= 0 
                  then error "no bytes in print_block" 
                  else printBlock (makeZeros min') min'
            first:rest -> do
              putStr $ show first ++ " "
              printBlock rest (min' - 1)
  
  let printAll n
        | n < actualMax = do
            let a = readBlock (bs device) n
            unless (isZeros a) $ do
              putStr $ show (fromIntegral n) ++ ": "
              printBlock a 8
              putStrLn ""
            printAll (n + 1)
        | otherwise = return ()
  
  printAll 0

p :: IO ()
p = printBlocks testDevice 0

-- Main function to run tests
main :: IO ()
main = do
  checkInvariant45
  format testDevice
  makeFiles testDevice sampleFiles1
  p